fendermint/vm/topdown/src/finality/null.rs

```diff
@@ -149,6 +132,41 @@ impl FinalityWithNull {
 
 /// All the private functions
 impl FinalityWithNull {
+    /// Get the first non-null block in the range [start, end].
+    fn min_nonnull_block(&self, start: BlockHeight, end: BlockHeight) -> Stm<Option<BlockHeight>> {
+        let cache = self.cached_data.read()?;
+        for h in start..=end {
+            if let Some(Some(_)) = cache.get_value(h) {
+                return Ok(Some(h));
+            }
+        }
+        Ok(None)
+    }
+
+    fn propose_next_height(&self) -> Stm<Option<BlockHeight>> {
+        let latest_height = if let Some(h) = self.latest_height()? {
+            h
+        } else {
+            tracing::debug!("no proposal yet as height not available");
+            return Ok(None);
+        };
+
+        let last_committed_height = if let Some(h) = self.last_committed_finality.read_clone()? {
+            h.height
+        } else {
+            unreachable!("last committed finality will be available at this point");
+        };
+        let next_proposal_height = last_committed_height + self.min_proposal_interval;
+
+        Ok(Some(if next_proposal_height < latest_height {
+            // safe to unwrap as we are sure `latest_height` will not be null block
+            self.min_nonnull_block(next_proposal_height, latest_height)?
```

Comment:
Looking forward here as you do is fine as long as the info in the cache will only be available once the receipt for that height has been executed and is available. If the cache can ensure this we can do this, if not we should look backward instead of forward to be safe.